llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,466 @@
+
+static llvm::SmallVector<MCPhysReg, 16> getPriorityReplacementRegs(
+    const Register Reg,
+    const llvm::SmallVector<MCPhysReg, 16> &ForbiddenPhysRegs,
+    const MachineFunction &MF, const VirtRegMap &VRM, LiveRegMatrix &LRM,
+    const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+    const AIEBaseRegisterInfo &TRI);
+
+static bool
+replaceReg(const Register Reg,
+           const llvm::SmallVector<MCPhysReg, 16> &ForbiddenPhysRegs,
+           llvm::SmallVector<MCPhysReg, 16> &AlreadyReplaced,
+           const MachineFunction &MF, VirtRegMap &VRM, LiveRegMatrix &LRM,
+           const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+           const AIEBaseRegisterInfo &TRI);
+
+static llvm::SmallVector<const MachineBasicBlock *, 4>
+getLoopBBs(const MachineLoopInfo &MLI);
+static bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                           llvm::SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                           llvm::SmallVector<MCPhysReg, 16> &UsedVirtRegs,
+                           VirtRegMap &VRM, MachineRegisterInfo &MRI);
+
+bool isCopyElimination(const MachineInstr &MI, const VirtRegMap &VRM);
+
+static llvm::SmallVector<MCPhysReg, 16>
+getBBForbiddenReplacements(const MachineBasicBlock *MBB, const VirtRegMap &VRM,
+                           const MachineRegisterInfo &MRI,
+                           const AIEBaseRegisterInfo *TRI);
+
+static bool
+isUndefReserved(const MachineOperand &Def, const MachineInstr &MI,
+                const MachineBasicBlock &MBB,
+                std::map<unsigned, llvm::SmallVector<const MachineInstr *, 4>>
+                    &UndefPartReuse,
+                const VirtRegMap &VRM);
+
+static bool isNotLastDef(const MachineOperand &Def,
+                         const MachineInstr &DefInstr,
+                         const MachineBasicBlock &MBB);
+
+static bool
+reuseSubRegs(llvm::SmallVector<const MachineInstr *, 4> &OrigInstr,
+             llvm::SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+static bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                                   const MachineInstr *InstrB);
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+  if (DisablePass)
+    return false;
+
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  auto &TRI =
+      *static_cast<const AIEBaseRegisterInfo *>(MRI.getTargetRegisterInfo());
+  VirtRegMap &VRM = getAnalysis<VirtRegMap>();
+  LiveRegMatrix &LRM = getAnalysis<LiveRegMatrix>();
+  LiveIntervals &LIS = getAnalysis<LiveIntervals>();
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+  bool Modified = false;
+
+  llvm::SmallVector<const MachineBasicBlock *, 4> LoopBBs = getLoopBBs(MLI);
+
+  if (LoopBBs.empty())
+    return false;
+
+  LLVM_DEBUG(llvm::dbgs() << "*** WAW Loop Register Rewriting: "
+                          << MF.getName());
+  LLVM_DEBUG(llvm::dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopBBs) {
```

Comment:
I think this loop can be refactored as a new function, like `renameMBBPhysRegs`.

Reply:
moved it